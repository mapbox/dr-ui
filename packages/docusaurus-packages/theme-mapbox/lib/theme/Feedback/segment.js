import forwardEvent from './forward-event';
import * as Sentry from '@sentry/browser';
import env from '@theme/AnalyticsShell/env';

// Send the event to Segment with forward event
export function sendToSegment({ state, props }) {
  const event = createSegmentEvent({ state, props });
  forwardEvent(event, props.webhook, (err) => {
    if (err) Sentry.captureException(err);
  });
}

// Creates event to send to Segment
function createSegmentEvent({ state, props }) {
  const { user, anonymousId, feedback, sessionId, category, categoryType } =
    state;
  const { site, section, location } = props;
  // set user if available else set anonymousId (needed for forward-event request)
  const identity = {
    ...(user && user.id ? { userId: user.id } : { anonymousId: anonymousId })
  };

  return {
    event: 'Sent docs feedback',
    ...identity,
    properties: {
      // the unique id for the feedbak session
      sessionId,
      // the feedback category that the user selected (helpful for yes, unhelpful for no)
      category,
      // the sub type for the category
      categoryType,
      // text feedback, if available
      ...(feedback && { feedback: feedback }),
      // name of current site (important for filtering in Mode)
      site,
      // (optional) name of section for longer pagers, helpful for fitering in Mode and identifying section areas
      section,
      // get page context
      page: location || undefined,
      // set user if available else set anonymousId (needed for Mode)
      userId: user && user.id ? user.id : undefined,
      anonymousId: anonymousId,
      // set plan, if available
      ...(user && user.plan && { plan: user.plan.id }),
      // get environment: staging or production
      environment: env()
    }
  };
}
