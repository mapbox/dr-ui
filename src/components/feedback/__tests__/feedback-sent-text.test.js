import React from 'react';
import Feedback from '..';
import toJson from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';

describe('Sent text feedback', () => {
  const feedback = mount(
    <Feedback
      site="dr-ui"
      location={{
        pathname: '/mapbox-gl-js/api/',
        hash: '#lnglat'
      }}
      webhook={{
        production: '',
        staging: ''
      }}
    />
  );
  feedback.setState({
    user: {
      id: 'crocsfan19',
      email: 'crocsfan19@mapbox.com',
      plan: {
        id: 'starter'
      }
    }
  });

  test('clicked yes and sent feedback', () => {
    // click yes
    feedback.setState({
      helpful: true
    });
    // simulate typing feedback
    feedback
      .find('#dr-ui--feedback-page-text')
      .at(0)
      .props()
      .onChange('cool beans!');
    // simulate submit button click
    feedback.find('#dr-ui--feedback-page-submit').at(0).simulate('click');

    expect(feedback.state()).toEqual({
      event: {
        event: 'Sent docs feedback',
        userId: 'crocsfan19',
        properties: {
          environment: 'staging',
          feedback: 'cool beans!',
          helpful: true,
          location: {
            hash: '',
            host: 'localhost',
            hostname: 'localhost',
            href: 'http://localhost/',
            origin: 'http://localhost',
            pathname: '/',
            search: ''
          },
          page: {
            hash: '#lnglat',
            pathname: '/mapbox-gl-js/api/'
          },
          plan: 'starter',
          section: undefined,
          site: 'dr-ui',
          userId: 'crocsfan19'
        }
      },
      feedback: 'cool beans!',
      feedbackSent: true,
      helpful: true,
      user: {
        id: 'crocsfan19',
        email: 'crocsfan19@mapbox.com',
        plan: {
          id: 'starter'
        }
      }
    });
  });
});

describe('Sent text feedback', () => {
  const feedback = shallow(
    <Feedback
      site="dr-ui"
      location={{
        pathname: '/mapbox-gl-js/api/',
        hash: '#lnglat'
      }}
    />
  );
  feedback.setState({
    helpful: true,
    feedback: 'Please add an example',
    feedbackSent: true
  });

  test('snapshot', () => {
    expect(toJson(feedback)).toMatchSnapshot();
  });
});
