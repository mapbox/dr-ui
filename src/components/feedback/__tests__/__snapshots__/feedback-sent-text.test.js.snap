// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sent text feedback snapshot 1`] = `
<div
  className="dr-ui--feedback"
>
  <AsideHeading>
    Thanks for your feedback.
  </AsideHeading>
</div>
`;
